draft-ietf-tls-esni.md

```diff
@@ -593,69 +629,73 @@ If the server sends an "encrypted_client_hello" extension, the client
-  to ESNI PSKs. ESNI PSKs offered by the client are associated with the ESNI
-  name. The server was unable to decrypt then ESNI name, so it should not resume
+  to ECHO PSKs. ECHO PSKs offered by the client are associated with the ECHO
+  name. The server was unable to decrypt then ECHO name, so it should not resume
   them when using the cleartext SNI name. This restriction allows a client to
   reject resumptions in {{auth-public-name}}.
 
-Note that an unrecognized ClientEncryptedSNI.record_digest value may be
-a GREASE ESNI extension (see {{grease-extensions}}), so it is necessary
+Note that an unrecognized ClientEncryptedCH.record_digest value may be
+a GREASE ECHO extension (see {{grease-extensions}}), so it is necessary
 for servers to proceed with the connection and rely on the client to abort if
-ESNI was required. In particular, the unrecognized value alone does not
-indicate a misconfigured ESNI advertisement ({{misconfiguration}}). Instead,
-servers can measure occurrences of the "esni_required" alert to detect this
+ECHO was required. In particular, the unrecognized value alone does not
+indicate a misconfigured ECHO advertisement ({{misconfiguration}}). Instead,
+servers can measure occurrences of the "echo_required" alert to detect this
 case.
 
-If the ClientEncryptedSNI value does match a known ESNIConfig, the server
-performs the following checks:
-
-- If the ClientEncryptedSNI.key_share group does not match one in the ESNIConfig.keys,
-  it MUST abort the connection with an "illegal_parameter" alert.
+If the ClientEncryptedCH value does match a known ECHOConfig, the server
+then decrypts ClientEncryptedCH.encrypted_ch, using the private key skR
+corresponding to ESNIConfig, as follows:
 
-Assuming these checks succeed, the server then computes K_sni
-and decrypts the ClientHelloInner value. If decryption fails, the server
-MUST abort the connection with a "decrypt_error" alert.
+~~~
+context = SetupBaseR(ClientEncryptedCH.enc, skR, "tls13-echo")
+ClientHelloInner = context.Open("", ClientEncryptedCH.encrypted_ch)
+echo_nonce = context.Export("tls13-echo-nonce", 16)
+echo_hrr_key = context.Export("tls13-echo-hrr-key", 16)
+~~~
 
-Once the ClientHelloInner has been decrypted, the server MUST
-scan it for any "outer_extension" extensions and substitute their
-values with the values in ClientHelloOuter. It MUST first verify that
-the hash found in the extension matches the hash of the extension
-to be interpolated in and if it does not, abort the connection
+If decryption fails, the server MUST abort the connection with
+a "decrypt_error" alert. Moreover, if there is no "echo_nonce"
+extension, or if its value does not match the derived echo_nonce,
+the server MUST abort the connection with a "decrypt_error" alert.
+Next, the server MUST scan ClientHelloInner for any "outer_extension"
+extensions and substitute their values with the values in ClientHelloOuter.
```

Comment:
(Note: this processing order is a little interesting because `echo_nonce` is parsed before `outer_extension` substitution, which means trying to parse an extension out of an invalid ClientHello. Not necessarily a problem, just mildly weird. But rejecting invalid things early is sensible enough.)